backend/LexBoxApi/Services/HgService.cs

```diff
@@ -262,6 +273,51 @@ public async Task<HttpContent> ExecuteHgRecover(string code, CancellationToken t
         return response;
     }
 
+    public Task<HttpContent> InvalidateDirCache(string code, CancellationToken token = default)
+    {
+        var repoPath = Path.Join(PrefixRepoFilePath(code));
+        if (Directory.Exists(repoPath))
+        {
+            // Invalidate NFS directory cache by forcing a write and re-read of the repo directory
+            var randomPath = Path.Join(repoPath, Path.GetRandomFileName());
+            while (File.Exists(randomPath) || Directory.Exists(randomPath)) { randomPath = Path.Join(repoPath, Path.GetRandomFileName()); }
+            try
+            {
+                // Create and delete a directory since that's slightly safer than a file
+                var d = Directory.CreateDirectory(randomPath);
+                d.Delete();
+            }
+            catch (Exception) { }
+        }
+        var result = ExecuteHgCommandServerCommand(code, "invalidatedircache", token);
+        return result;
+    }
+
+    public async Task<string> GetTipHash(string code, CancellationToken token = default)
+    {
+        var content = await ExecuteHgCommandServerCommand(code, "tip", token);
+        return await content.ReadAsStringAsync();
+    }
+
+    private async Task WaitForRepoEmptyState(string code, RepoEmptyState expectedState, int timeoutMs = 30_000, CancellationToken token = default)
+    {
+        // Set timeout so unforeseen errors can't cause an infinite loop
+        var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+        timeoutSource.CancelAfter(30_000);
+        var done = false;
+        while (!done && !timeoutSource.IsCancellationRequested)
+        {
+            var hash = await GetTipHash(code, timeoutSource.Token);
+            var isEmpty = hash == AllZeroHash;
+            done = expectedState switch
+            {
+                RepoEmptyState.Empty => isEmpty,
+                RepoEmptyState.NonEmpty => !isEmpty
+            };
+            if (!done) await Task.Delay(2500);
```

Comment:
should pass the cancelation token into Delay otherwise the timeout will trigger but this delay will still take place

Reply:
Done in commit 18ae58e7, I'd missed that Task.Delay can take a cancellation token.